versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java

```diff
@@ -828,25 +852,48 @@ protected Hash casOpLoop(
         boolean casSuccess;
         switch (opVariant) {
           case CREATE_REF:
-            casSuccess = createNamedReference(ctx, ref, result.newHead);
+            ReferenceCreatedResult refCreatedResult = ((ReferenceCreatedResult) result.result);
+            casSuccess = createNamedReference(ctx, ref, refCreatedResult.getHash());
             break;
           case DELETE_REF:
             casSuccess = deleteNamedReference(ctx, ref, refHead.getRef());
             break;
           case REF_UPDATE:
+            ReferenceAssignedResult refAssignedResult = (ReferenceAssignedResult) result.result;
+            if (refHead.getRef().getHash().equals(refAssignedResult.getCurrentHash().asBytes())) {
+              // Target branch HEAD did not change
+              return tryState.success(result.result);
+            }
+            casSuccess =
+                updateNamedReference(
+                    ctx, ref, refHead.getRef(), refAssignedResult.getCurrentHash());
+            break;
           case COMMIT:
-            if (refHead.getRef().getHash().equals(result.newHead.asBytes())) {
+            CommitResult<?> commitResult = (CommitResult<?>) result.result;
+            if (refHead.getRef().getHash().equals(commitResult.getCommitHash().asBytes())) {
+              // Target branch HEAD did not change
+              return tryState.success(result.result);
+            }
+            casSuccess =
+                updateNamedReference(ctx, ref, refHead.getRef(), commitResult.getCommitHash());
+            break;
+          case MERGE:
+            MergeResult<?> mergeResult = ((MergeResult<?>) result.result);
+            Hash newHash = Objects.requireNonNull(mergeResult.getResultantTargetHash());
+            if (refHead.getRef().getHash().equals(newHash.asBytes())) {
               // Target branch HEAD did not change
-              return tryState.success(result.newHead);
+              return tryState.success(result.result);
             }
-            casSuccess = updateNamedReference(ctx, ref, refHead.getRef(), result.newHead);
+            casSuccess =
+                updateNamedReference(
+                    ctx, ref, refHead.getRef(), mergeResult.getResultantTargetHash());
             break;
           default:
             throw new UnsupportedOperationException("Unknown opVariant " + opVariant);
         }
 
         if (!casSuccess) {
-          if (opVariant == CasOpVariant.COMMIT) {
+          if (opVariant == CasOpVariant.COMMIT || opVariant == CasOpVariant.MERGE) {
             cleanUpCommitCas(ctx, individualCommits, individualKeyLists);
```

Comment:
Just noting (nothing to change here) - only for posterity why this is _not_ in the new model:
Initially I ported this behavior to the new storage model as well, but (manual, bad "heavy load") tests proved that this can go sideways - up to the point that two concurrent writers produce the exact same commit, one wins and bumps HEAD, the other one "loses" and deletes the commit -> HEAD points to a non-existing commit. It's a rather artificial and will probably never happen in real-live, but still, there's a chance.